File: maximum-subarray/maximum-subarray.java
class Solution {

    int max(int a,int b)
    {
        if(a>b)
        return a;
        else
        return b;
    }
    
    public int maxSubArray(int[] nums) {
        
    //step  1:
    int sum = 0;
    int maxi = -100000;

    int n = nums.length;
    for(int i=0;i<n;i++)
    {
        sum = sum + nums[i];
        maxi  = max(sum,maxi);

        if(sum < 0)
        {
            sum = 0;
        }
    }
    return maxi;
    }
}

File: maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) 
    {
        // step 1 : I have initialized two variables...
        int sum = 0;
        int maxSum = INT_MIN;
        //step 2:
        int n = nums.size();
        for(int i=0;i<n;i++)
        {
                sum = sum + nums[i]; // 0 + (-1)  = -1
                maxSum  = max(sum,maxSum); // -1  -infinity  --> //-1  
                // OR
                // if(maxSum < sum)
                // {
                //     maxSum = sum;
                // }

                if(sum < 0)
                {
                    sum = 0;  //sum  = 0 
                }
        } 
        return maxSum;
    }
};